### PR TITLE
Update Objective-C check for Mathematica

### LINKED ISSUE #2197: Objective-C is being confused with Mathematica
[This file](https://github.com/d235j/360Controller/blob/master/360Daemon/360Daemon.m) is actually Objective-C, but the Linguist thinks it's Mathematica. There are no other files that I know of that are like this.

### PR DESCRIPTION
**Pull request by phase**
Fixes #2197 Code from @pchaigno, I just made the PR. :rabbit:

### CODE REVIEW

**Comment by phase** [comment]
```objective-c
@interface StyleViewController : TTViewController {
@private
  TTStyle*  _style;
  TTStyle*  _styleHighlight;
  TTStyle*  _styleDisabled;
  TTStyle*  _styleSelected;
  NSString* _styleType;
}
@end
```
This fails. Can we separate the `.h` & `.m`? I see no reason for them to be shared in one variable.

**Comment by phase** [comment]
The `#include` in `store.m` is throwing an error because it has a string of C code. I wonder if we have to have `include`, as `import|define` might be enough.

**Comment by pchaigno** [comment]
> The #include in store.m is throwing an error because it has a string of C code.

We could exclude it from the test.
<br/>

> I wonder if we have to have include, as import|define might be enough.

You should check against a corpus of files. If you have a link to a GitHub search, @arfon can send you the first 1,000 file results.
<br/>

> We may also be able to check if the code is in a string, or make string-checking a standard for all languages.

There was [a GSoC idea](https://github.com/github/gsoc/blob/040f3f2c294f9e4f5bd8e1ab2269f96e23bb0fe8/README.md#use-grammar-bundles-to-tokenize-languages) related to that.

**Comment by phase** [comment]
If we exclude one file, what if another file comes along somewhere with the same code? We can't just ignore every file that has a string of another language.

File: lib/linguist/heuristics.rb [hunk 1]
```diff
@@ -61,9 +61,6 @@ def call(data)
       @heuristic.call(data)
     end
 
-    # Common heuristics
-    ObjectiveCRegex = /^[ \t]*@(interface|class|protocol|property|end|synchronised|selector|implementation)\b/
```
Comment (arfon) [inline lib/linguist/heuristics.rb]: I'm not sure I understand the rationale for having a different heuristic for `Objective-C` in different scenarios. Is this necessary? There are good reasons for wanting to extract this into a variable.

Comment (phase) [inline lib/linguist/heuristics.rb]: There are two different ways for detecting ObjC, each one for a different file type.

`.h` will be something like a class, so it has the check at line 81

`.m` will include something, so there's that.

Comment (pchaigno) [inline lib/linguist/heuristics.rb]: It much more difficult to differentiate Objective-C from C and C++ than it is to differentiate it from Limbo, M, MUF, Mathematica, Matlab and Mercury. That's why I think we can use a simpler regex (faster?) for the second case.

**Comment by arfon** [comment]
> You should check against a corpus of files. If you have a link to a GitHub search, @arfon can send you the first 1,000 file results.

Very happy to do this. If you can construct a search [here](https://github.com/search/advanced) @phase then I can go and grab a whole collection of files to test against.

**Comment by arfon** [comment]
:flags: flagging this as stale :flags:

**Pull request closed without merging**